ecommerce/ordering/lib/ordering/order.rb

```diff
@@ -79,35 +79,34 @@ def cancel
     end
 
     on ItemAddedToBasket do |event|
-      product_id = event.data[:product_id]
-      order_line = find_order_line(product_id)
-      unless order_line
-        order_line = create_order_line(product_id)
-        @order_lines << order_line
-      end
-      order_line.increase_quantity
+      @basket.increase_quantity(event.data[:product_id])
     end
 
     on ItemRemovedFromBasket do |event|
-      product_id = event.data[:product_id]
-      order_line = find_order_line(product_id)
-      return unless order_line
-      order_line.decrease_quantity
-      remove_order_line(order_line) if order_line.empty?
+      @basket.decrease_quantity(event.data[:product_id])
+    end
+  end
+
+  class Basket
+    def initialize
+      @order_lines = {}
```

Comment:
How about:

```suggestion
      @order_lines = Hash.new { |basket, product_id| basket[product_id] = 0 }
```

and you don't need `fetch(..., 0)` anymore